### PR TITLE
Object metadata should be optional in resumable uploads #332

### PR DESCRIPTION
**Pull request by martinzink**
We are working on google-cloud integration for our project [Apache Nifi Minifi CPP](https://github.com/apache/nifi-minifi-cpp) and fake-gcs-server seems to be an awesome way to create integration tests.
However I encountered some issues, which made impossible to integrate fake-gcs-server with a C++ app

I already fixed the issues from google-cloud-cpp's side
https://github.com/googleapis/google-cloud-cpp/pull/8446
https://github.com/googleapis/google-cloud-cpp/pull/8477

The last issue that blocks using fake-gcs-server with a cpp application seems to be #332 
As described in the issue fake-gcs-server is stricter than google cloud storage when it comes to resumable upload requests.
With this PR the object metadata is optional when creating a resumable upload request.

I also added cpp examples so it can be verified that the resumable uploads are working now with fake-gcs-server.

### CODE REVIEW

File: examples/cpp/list_bucket.cpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#include "google/cloud/storage/client.h"
```
Comment (fsouza) [inline examples/cpp/list_bucket.cpp:1]: Can you add this to CI? 

For that we need:

1. A script in `./ci`
2. A docker image where that script can run (probably something with cmake preinstalled?)
3. Then add it to the matrix here: https://github.com/fsouza/fake-gcs-server/blob/cf3fcb083e19553636419818e29f84825bd1e13c/.github/workflows/main.yml#L120-L139

File: fakestorage/upload.go [hunk 2]
```diff
@@ -327,10 +327,7 @@ func (s *Server) multipartUpload(bucketName string, r *http.Request) jsonRespons
 func (s *Server) resumableUpload(bucketName string, r *http.Request) jsonResponse {
 	predefinedACL := r.URL.Query().Get("predefinedAcl")
 	contentEncoding := r.URL.Query().Get("contentEncoding")
-	metadata, err := loadMetadata(r.Body)
-	if err != nil {
```
Comment (gaul) [inline fakestorage/upload.go:331]: Can you add a test that demonstrates this behavior?

Also is it possible to retain this check if we narrow it to ignore specific kinds of errors, e.g., missing metadata?

File: examples/cpp/list_bucket.cpp (hunk 1 above)
Comment (fsouza) [reply examples/cpp/list_bucket.cpp]: Oh 3 minutes isn't too bad given that CI runs things in parallel and running the tests on Windows takes 6 minutes anyway. We could definitely optimize it later though (would probably require maintaining a custom image, or leverage some caching mechanism in GitHub Actions).

**Review by fsouza — APPROVED** [review]
Thanks for contributing and sorry for the delayed review!

**Pull request merged**